src/MgvOfferTaking.sol

```diff
@@ -188,21 +194,31 @@ abstract contract MgvOfferTaking is MgvHasOffers {
 
   /* ## Internal market order */
   //+clear+
-  /* `internalMarketOrder` works recursively. Going downward, each successive offer is executed until the market order stops (due to: volume exhausted, bad price, or empty book). Then the [reentrancy lock is lifted](#internalMarketOrder/liftReentrancy). Going upward, each offer's `maker` contract is called again with its remaining gas and given the chance to update its offers on the book.
-
-    The last argument is a boolean named `proceed`. If an offer was not executed, it means the price has become too high. In that case, we notify the next recursive call that the market order should end. In this initial call, no offer has been executed yet so `proceed` is true. */
-  function internalMarketOrder(Pair storage pair, MultiOrder memory mor, MgvLib.SingleOrder memory sor, bool proceed)
-    internal
-  {
+  /* `internalMarketOrder` works recursively. Going downward, each successive offer is executed until the market order stops (due to: volume exhausted, bad price, or empty book). Then the [reentrancy lock is lifted](#internalMarketOrder/liftReentrancy). Going upward, each offer's `maker` contract is called again with its remaining gas and given the chance to update its offers on the book. */
+  function internalMarketOrder(Pair storage pair, MultiOrder memory mor, MgvLib.SingleOrder memory sor) internal {
     unchecked {
+      if (mor.fillVolume == 0 || sor.offerId == 0 || mor.maxDepth == 0 || !sor.offer.tick().better(mor.maxTick)) {
+        return endInternalMarketOrder(pair, mor, sor);
+      }
+
+      /* Load additional information about the offer. We don't do it earlier to save one storage read in case `proceed` was false. */
+      sor.offerDetail = pair.offerData[sor.offerId].detail;
+
+      mor.gasreqLeft += sor.offerDetail.gasreq();
+
+      if (gasleft() < mor.gasreqLeft + sor.local.offer_gasbase()) {
+        return endInternalMarketOrder(pair, mor, sor);
+      }
```

Comment:
I like how this simplifies this method and makes it clear what the end-conditions are and how they are handled.

Comments need to be updated, but we can do that when/if we decide to move forward with this.